### PR TITLE
Include region extraction from subsets

### PR DESCRIPTION
**Pull request by nmearl**
This PR addresses the need to get glue subsets as astropy region objects from generic viewers.

```python
>>> subsets = app.get_subsets_from_viewer('spectrum-viewer')
>>> subsets
{'Subset 1': <RectanglePixelRegion(PixCoord(x=4.3897305938417555e-07, y=2281.5), width=5.972716083637954e-08, height=4563, angle=0.0 deg)>}
```

### CODE REVIEW

**Comment by eteq** [comment]
Question: @nmearl, is this meant as an intermediate step towards getting out `SpectralRegion` objects?  Or were you thinking this is the end API for users? The concept notebook highlights why it's important to have an "easy" way to get out a `SpectralRegion` as opposed to an astropy region, although now that you show this I can see the argument for a generic region-based API and a specviz *specialization* to `SpectralRegion`

**Comment by nmearl** [comment]
Your latter point is correct: this is the generic api that is used in jdaviz. There will be another PR that's based on Duy's specviz implementation that provided a specific method that takes the astropy regions provided by this method and converts them to `SpectralRegions`.

File: notebooks/Example.ipynb [hunk 1]
```diff
@@ -30,6 +175,125 @@
...
+    "from glue.core.data import Data\n",
+    "import numpy as np\n",
+    "\n",
+    "from glue.core.roi import RectangularROI, PolygonalROI, CircularROI, \\\n",
+    "    PointROI, XRangeROI, YRangeROI, AbstractMplRoi\n",
+    "\n",
+    "from glue.core.subset import RoiSubsetState, RangeSubsetState, OrState, \\\n",
+    "    AndState, XorState, MultiOrState, MultiRangeSubsetState\n",
+    "\n",
+    "# data = Data(flux=np.ones((256, 128, 128)), label='Test Flux2')\n",
```
Comment (rosteen) [inline notebooks/Example.ipynb:217]:
```suggestion
    "data = Data(flux=np.ones((256, 128, 128)), label='Test Flux2')\n",
```

This cell fails with this line commented out, should either uncomment it or drop all this subset testing from the generic example notebook.

Comment (nmearl) [reply notebooks/Example.ipynb:217]: Whoops! Those changes should definitely not be in there. Thanks for catching that, @rosteen!

File: jdaviz/app.py [hunk 2]
```diff
@@ -255,8 +256,8 @@ def get_viewer(self, viewer_reference):
         """
         return self._viewer_by_reference(viewer_reference)
 
-    def get_data_from_viewer(self, viewer_reference, data_label=None, cls=None,
-                             include_subsets=True):
+    def get_data_from_viewer(self, viewer_reference, data_label='None',
+                             cls='default', include_subsets=True):
```
Comment (eteq) [inline jdaviz/app.py:260]: The docstring should be updated to reflect the new role of `default`

File: jdaviz/app.py [hunk 3]
```diff
@@ -342,6 +343,102 @@ def get_data_from_viewer(self, viewer_reference, data_label=None, cls=None,
...
+        ostensibly just a view into a 3D data set, it is necessary to first
+        reduce the dimensions of the data, then retrieve the subset information
+        as a regions object. This means that the returned y extents in the
+        region are not strictly representative of the subset range in y.
+
+        Parameters
+        ----------
+        viewer_reference : str
+            The reference to the viewer defined with the ``reference`` key
+            in the yaml configuration file.
```
Comment (eteq) [inline jdaviz/app.py:363]: For a user-level API this is pretty opaque - the user needs to be able to get these regions without looking at a configuration file. I realize this is used elsewhere though, so I think this is more of a separate issue and I think it neednt hold up this particular PR.

File: jdaviz/app.py [hunk 4]
```diff
@@ -342,6 +343,102 @@ def get_data_from_viewer(self, viewer_reference, data_label=None, cls=None,
...
+        data_label : str, optional
+            Optionally provide a label to retrieve a specific data set from the
+            viewer instance.
+
+        Returns
+        -------
+        data : dict
+            A dict of the transformed Glue subset objects, with keys
+            representing the subset name.
```
Comment (eteq) [inline jdaviz/app.py:372]:
```suggestion
        data : dict
            A dict of the transformed Glue subset objects, with keys
            representing the subset name and values as astropy regions
            objects.
```

File: jdaviz/app.py (hunk 3 above)
Comment (eteq) [reply jdaviz/app.py:363]: (see #175)

**Comment by eteq** [comment]
(did a few quick fixes and rebase to allow merge)

File: jdaviz/app.py [hunk 5]
```diff
@@ -311,7 +318,10 @@ def get_data_from_viewer(self, viewer_reference, data_label=None, cls=None,
             corresponding viewer data labels.
         """
         viewer = self.get_viewer(viewer_reference)
-        cls = cls or viewer.default_class
+        cls = viewer.default_class if cls == 'default' else cls
+        if not isclass(cls):
```
Comment (nmearl) [inline jdaviz/app.py:322]: This is not correct: `cls` can be a class object, 'default' or `None` (and in fact `None` is necessary case, hence the failing tests.)

**Pull request merged**